### PR TITLE
Enable collection during allocation

### PR DESCRIPTION
**Pull request by dwightguth**
We add a call to koreCollect in `arenaAlloc` if we are close to running out of free space, and only allocate a new block if we're still out of space after the collection. We also modify the behavior of the alwaysgcspace so that it is cleared after each rewrite step.

This ought to ensure that there is never a long period of time where collection cannot occur, since collection can occur during any allocation with the exception of some allocations from inside hooks. It also makes it much less likely that the application will allocate more blocks than are strictly needed, thus reducing memory usage. Additionally, since we are no longer periodically making an expensive check about whether or not to perform a collection, there is a slight improvement in runtime performance.

Limitations: we still only collect things allocated via koreAllocAlwaysGC after each rewrite step. This needs to be improved in future PRs.

### CODE REVIEW

File: runtime/collect/StackMapParser.cpp [hunk 1]
```diff
@@ -39,26 +39,35 @@ void parseStackMap() {
       void *ip = (void *)(func.FunctionAddress + InstructionOffset);
       int32_t NumDeopts = *(int32_t *)(stackMapRecord + 24 + 2 * 12);
       uint16_t RelocationOffset = 0;
-      for (uint16_t j = 4 + NumDeopts; j < NumLocations; j += 2) {
-        uint8_t type = *(uint8_t *)(stackMapRecord + 16 + j * 12);
-        if (type == 5) {
+      for (uint16_t j = 3 + NumDeopts; j < NumLocations; j += 2) {
```
Comment (Baltoli) [inline runtime/collect/StackMapParser.cpp:42]: What part of the schema is this looking at, and what's the difference between offset 4 and offset 3? Some further comments tying this parser to [the schema](https://llvm.org/docs/StackMaps.html#stack-map-format) would be nice - I have no doubt it's correct, but it's difficult to follow the byte arithmetic to make sure it's correct.

File: runtime/collect/StackMapParser.cpp [hunk 2]
```diff
@@ -39,26 +39,35 @@ void parseStackMap() {
...
           // thus do not need to be relocated.
           RelocationOffset++;
           continue;
         }
-        if (type != 3) {
+        if (base_type != 3) {
           abort();
         }
-        int32_t offset = *(int32_t *)(stackMapRecord + 24 + j * 12);
+        uint8_t derived_type = *(uint8_t *)(stackMapRecord + 16 + (j + 1) * 12);
```
Comment (Baltoli) [inline runtime/collect/StackMapParser.cpp:55]: For example, as above this kind of thing is difficult to grok - a comment mapping the pointer computation to a path into the schema would be good (e.g. for the sake of demonstration on a contrived example not related to this particular line, something like `StkMapRecord[i].Location[j].DwarfRegNum` would clarify exactly what bytes we're expecting to access.

There is a bit of an additional burden involved with keeping this in sync, I guess - how often is this code likely to be modified? If it's only once in a while, then probably worth the effort of keeping the comments.

**Review by Baltoli — CHANGES_REQUESTED** [review]

File: runtime/collect/StackMapParser.cpp (hunk 2 above)
Comment (dwightguth) [reply runtime/collect/StackMapParser.cpp:55]: Good call about the comments. I'll go ahead and go through this function and add them.

File: lib/codegen/Decision.cpp [hunk 3]
```diff
@@ -888,23 +887,19 @@ void makeStepFunction(
...
   if (search) {
     auto result = new llvm::LoadInst(bufType, resultBuffer, "", stuck);
     llvm::ReturnInst::Create(module->getContext(), result, stuck);
   } else {
-    auto phi
-        = llvm::PHINode::Create(collectedVal->getType(), 2, "phi_1", stuck);
+    auto phi = llvm::PHINode::Create(val->getType(), 2, "phi_1", stuck);
     phi->addIncoming(val, block);
-    phi->addIncoming(collectedVal, pre_stuck);
+    phi->addIncoming(val, pre_stuck);
```
[the remark this answers is not in the document]
Comment (dwightguth) [reply lib/codegen/Decision.cpp:902]: Not really. It was just a lazy update to the function. I have taken this phi node out and it seems to still pass.

**Review by Baltoli — APPROVED** [review]
Great - the comments on the stack map parser are really helpful for understanding what's going on.

**Review by theo25 — APPROVED** [review]
This looks good to me.

**Review by theo25 — APPROVED** [review]

File: lib/llvm/EmitGCLayoutInfo.cpp [hunk 4]
```diff
@@ -64,7 +64,7 @@ struct EmitGCLayoutInfo : public ModulePass {
               if (!Ty->isStructTy()) {
                 error(Ty);
               }
-              auto StructTy = llvm::cast<StructType>(Ty);
+              auto StructTy = cast<StructType>(Ty);
               if (!StructTy->hasName()) {
                 error(StructTy);
               }
```
Comment (Baltoli) [inline lib/llvm/EmitGCLayoutInfo.cpp:70]: This whole stanza could be a bit more succinct:

```c++
auto StructTy = dyn_cast<StructType>(Ty);
if (!StructTy || !StructTy->hasName()) {
  error(StructTy);
}
```

**Review by Baltoli — APPROVED** [review]
One small change, but otherwise looks good to me.

**Pull request merged**